spike_main/spike.cc

```diff
@@ -128,6 +128,23 @@ static bool check_mem_overlap(const mem_cfg_t& L, const mem_cfg_t& R)
   return std::max(L.get_base(), R.get_base()) <= std::min(L.get_inclusive_end(), R.get_inclusive_end());
 }
 
+// there is a weird corner-case preventing two memory regions to be merged -
+// this can happen when the resulting size of a region is 2^64.
+// such regions are not representable with mem_cfg_t class
+static bool check_if_merge_covers_64bit_space(const mem_cfg_t& L,
+                                              const mem_cfg_t& R)
+{
+  if (!check_mem_overlap(L, R))
+    return false;
+
+  reg_t start = std::min(L.get_base(), R.get_base());
+  reg_t end = std::max(L.get_inclusive_end(), R.get_inclusive_end());
+
+  // if [base == 0] and [inclusive_end == (2 ^ 64 - 1)] - we can't merge such
+  // regions, since the mem_cfg_t does not support the resulting size
+  return (start == 0ull) && (end == std::numeric_limits<uint64_t>::max());
```

Comment:
`mem_cfg_t` uses `reg_t`, not `uint64_t`.

In C++, is there a way to get the type of a method at compile time? I.e. instead of `uint64_t` or `reg_t` here, you use the return type of `get_inclusive_end()`.

Reply:
> In C++, is there a way to get the type of a method at compile time? 
I've replaced  explicit usage of `reg_t` type with an auto placeholder - this should be enough for `start` / `end` calculations.

I still use uint64_t type explicitly, since this type is exactly what we need if we want to get the maximum value for  **64** bit number.